# Limpeza e Tratamento dos Dados

Nesse notebook iremos fazer a limpeza e o tratamento do [dataset](https://www.kaggle.com/datasets/carloseduardogo/apartment-prices-in-goinia-gois-brazil) que iremos usar nesse projeto. Após o processo de limpeza e tratamento, iremos salvar um novo dataset no formato parquet.

## Importando as Bibliotecas

Primeiro, vamos importar as bibliotecas que usaremos nessa etapa.

In [497]:
import numpy as np
import pandas as pd

## Carregando o Dataset

Agora, vamos carregar o dataset e ver algumas informações gerais.

In [498]:
df = pd.read_csv("../data/raw/dataset.csv", sep="|")

C:\Users\omath\AppData\Local\Temp\ipykernel_6772\3474246224.py:1: DtypeWarning: Columns (0: freezer) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../data/raw/dataset.csv", sep="|")


In [499]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9839 entries, 0 to 9838
Data columns (total 79 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   id                      9839 non-null   object
 1   tag_card                3610 non-null   str   
 2   business_type           9839 non-null   str   
 3   price                   9839 non-null   str   
 4   condo_fee               9768 non-null   str   
 5   iptu                    9768 non-null   str   
 6   address                 9839 non-null   str   
 7   floorSize               9839 non-null   str   
 8   numberOfRooms           9837 non-null   str   
 9   numberOfBathroomsTotal  9837 non-null   str   
 10  numberOfParkingSpaces   9573 non-null   str   
 11  floorLevel              4494 non-null   str   
 12  pool                    4946 non-null   str   
 13  elevator                4984 non-null   str   
 14  barbecueGrill           4580 non-null   str   
 15  gatedCommunity 

## Renomeando as Features

Vamos renomear cada uma das features para manter os nomes em português e para garantir melhor entendimento do dataset.

In [500]:
df = df.rename({
    "id": "id",
    "tag_card": "anuncio",
    "business_type": "negocio",
    "price": "preco",
    "condo_fee": "condominio",
    "iptu": "iptu",
    "address": "endereco",
    "floorSize": "tamanho",
    "numberOfRooms": "quartos",
    "numberOfBathroomsTotal": "banheiros",
    "numberOfParkingSpaces": "vagas_estacionamento",
    "floorLevel": "andar",
    "pool": "piscina",
    "elevator": "elevador",
    "barbecueGrill": "churrasqueira",
    "gatedCommunity": "condominio_fechado",
    "gym": "academia",
    "gourmetSpace": "espaco_gourmet",
    "garden": "jardim",
    "playground": "parque_infantil",
    "partyHall": "salao_festas",
    "laundry": "lavanderia",
    "disabledAccess": "acesso_pcd",
    "coworking": "espaco_coworking",
    "electronicGate": "portao_eletronica",
    "sauna": "sauna",
    "spa": "spa",
    "complete_description": "descricao",
    "petsAllowed": "permitido_animais",
    "airConditioning": "ar_condicionado",
    "gourmetBalcony": "varanda_gourmet",
    "sportsCourt": "quadra_poliesportiva",
    "bicyclesPlace": "bicicletario",
    "concierge24h": "portaria_24h",
    "americanKitchen": "cozinha_americana",
    "balcony": "sacada",
    "intercom": "interfone",
    "serviceArea": "area_servico",
    "largeKitchen": "cozinha_grande",
    "furnished": "mobiliado",
    "kitchen": "cozinha",
    "adultGameRoom": "salao_jogos",
    "deposit": "deposito",
    "watchman": "porteiro",
    "security24Hours": "seguranca_24h",
    "bedroomWardrobe": "armario_embutido",
    "internetAccess": "internet",
    "closet": "armario",
    "gourmetKitchen": "cozinha_gourmet",
    "electricGenerator": "gerador_eletrico",
    "alarmSystem": "alarmes",
    "toysPlace": "brinquedoteca",
    "cinema": "cinema",
    "heating": "aquecimento",
    "builtinWardrobe": "moveis_embutidos",
    "kitchenCabinets": "armarios_cozinha",
    "safetyCircuit": "circuito_seguranca",
    "garage": "garagem",
    "tennisCourt": "quadra_tenis",
    "greenSpace": "area_verde",
    "squash": "squash",
    "fireplace": "lareira",
    "homeOffice": "escritorio",
    "bathroomCabinets": "armarios_banheiro",
    "gasShower": "chuveiro_gas",
    "blindexBox": "box_blindex",
    "reception": "recepcao",
    "cableTv": "tv_cabo",
    "bathtub": "banheira",
    "privatePool": "piscina_privativa",
    "backyard": "quintal",
    "grass": "gramado",
    "largeWindow": "janelas_grandes",
    "naturalVentilation": "ventilacao_natural",
    "cooker": "fogao",
    "hikingTrail": "trilha_caminhada",
    "panoramicView": "vista_panoramica",
    "massage": "massagem",
    "freezer": "freezer",
}, axis="columns")

## Tratando e Limpando as Features

Agora, vamos analisar feature por feature, tratar e limpar seus valores.

In [501]:
def get_info(column, values=True):
    num_na = column.isna().sum()
    num_total = len(column)

    print(f"dtype: {column.dtype}")
    print(f"NA: {num_na}/{num_total} ({100*num_na/num_total:.2f}%)")

    counts = column.value_counts(dropna=False)

    if values:
        df_values = pd.DataFrame({
            "contagem": counts,
            "porcentagem": 100*counts/num_total
        })

        df_values["porcentagem"] = (df_values["porcentagem"]
            .map(lambda x: f"{x:.2f} %")
        )
        
        print(f"valores:\n{df_values}")

### Feature `id`

In [502]:
get_info(df["id"])

dtype: object
NA: 0/9839 (0.00%)
valores:
                                         contagem porcentagem
id                                                           
62750619205425081544300816117885861273          1      0.01 %
302678336848826931249389939487192177877         1      0.01 %
130359292211767838527233708900928593543         1      0.01 %
191007733490619722070589252030900058169         1      0.01 %
200978065182550317495009592604134775241         1      0.01 %
...                                           ...         ...
52951181712583241391100308271868718092          1      0.01 %
257750376693945269556421609451713199731         1      0.01 %
217594899359609635550616912254590827557         1      0.01 %
32608285309092824311552237605111230321          1      0.01 %
73165821369918148679896241496674788865          1      0.01 %

[9839 rows x 2 columns]


Como podemos ver, a feature apenas representa os id's dos imóveis. Como essa informação não estará presente durante a inferência, então podemos remover ela.

In [503]:
df = df.drop("id", axis="columns")

Antes de prosseguir, vamos investigar rapidamente a existência de duplicatas e removê-las.

In [504]:
df[df.duplicated()].info()

<class 'pandas.DataFrame'>
Index: 2583 entries, 37 to 9827
Data columns (total 78 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   anuncio               1071 non-null   str  
 1   negocio               2583 non-null   str  
 2   preco                 2583 non-null   str  
 3   condominio            2554 non-null   str  
 4   iptu                  2554 non-null   str  
 5   endereco              2583 non-null   str  
 6   tamanho               2583 non-null   str  
 7   quartos               2582 non-null   str  
 8   banheiros             2582 non-null   str  
 9   vagas_estacionamento  2519 non-null   str  
 10  andar                 1232 non-null   str  
 11  piscina               1235 non-null   str  
 12  elevador              1253 non-null   str  
 13  churrasqueira         1105 non-null   str  
 14  condominio_fechado    985 non-null    str  
 15  academia              1263 non-null   str  
 16  espaco_gourmet       

Como podemos ver, existem $2583$ registros duplicados. Vamos então remover eles para garantir a integridade dos nossos dados.

In [505]:
df = df.drop_duplicates()

In [506]:
df.info()

<class 'pandas.DataFrame'>
Index: 7256 entries, 0 to 9838
Data columns (total 78 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   anuncio               2539 non-null   str  
 1   negocio               7256 non-null   str  
 2   preco                 7256 non-null   str  
 3   condominio            7214 non-null   str  
 4   iptu                  7214 non-null   str  
 5   endereco              7256 non-null   str  
 6   tamanho               7256 non-null   str  
 7   quartos               7255 non-null   str  
 8   banheiros             7255 non-null   str  
 9   vagas_estacionamento  7054 non-null   str  
 10  andar                 3262 non-null   str  
 11  piscina               3711 non-null   str  
 12  elevador              3731 non-null   str  
 13  churrasqueira         3475 non-null   str  
 14  condominio_fechado    2524 non-null   str  
 15  academia              3761 non-null   str  
 16  espaco_gourmet        

### Feature `anuncio`

In [507]:
get_info(df["anuncio"])

dtype: str
NA: 4717/7256 (65.01%)
valores:
                   contagem porcentagem
anuncio                                
NaN                    4717     65.01 %
Destaque               2331     32.13 %
Super Destaque          166      2.29 %
Em construção            24      0.33 %
Pronto para morar        14      0.19 %
Na planta                 4      0.06 %


Vamos remover essa feature pois, assim como `id`, não teremos acesso a ela durante a etapa de inferência.

In [508]:
df = df.drop("anuncio", axis="columns")

### Feature `negocio`

In [509]:
get_info(df["negocio"])

dtype: str
NA: 0/7256 (0.00%)
valores:
             contagem porcentagem
negocio                          
Venda            7214     99.42 %
A partir de        39      0.54 %
Endereço            3      0.04 %


Como podemos ver, essa feature tem variância baixíssima e assim como ocorreu com `id` e `anuncio` não teremos acesso a essa feature durante a inferência. Portanto podemos removê-la também.

In [510]:
df = df.drop("negocio", axis="columns")

### Feature `preco` (target)

In [511]:
get_info(df["preco"])

dtype: str
NA: 0/7256 (0.00%)
valores:
              contagem porcentagem
preco                             
R$ 450.000         145      2.00 %
R$ 550.000         123      1.70 %
R$ 650.000         114      1.57 %
R$ 350.000          96      1.32 %
R$ 850.000          78      1.07 %
...                ...         ...
R$ 713.000           1      0.01 %
R$ 1.280.007         1      0.01 %
R$ 533.735           1      0.01 %
R$ 372.209           1      0.01 %
R$ 494.000           1      0.01 %

[1785 rows x 2 columns]


Pelas informações acima podemos notar que o preço está codificado como texto com um padrão simples. Vamos checar se toda a coluna segue esse padrão.

In [512]:
df[~df["preco"].str.match(r"^R[$] (\d+\.)*(\d+)+(,\d+)?")]["preco"]

51     Sob consulta
159    Sob consulta
325    Sob consulta
Name: preco, dtype: str

Como podemos ver alguns valores não seguem o padrão e também não correspondem a um preço válido. Como não há valores faltantes então vamos eliminar esses registros que não seguem o padrão já que não podemos usâ-los para o treinamento.

In [513]:
df = df[df["preco"].str.match(r"^R[$] (\d+\.)*(\d+)+(,\d+)?")]

Agora vamos extrair os valores numéricos e converter a coluna para um tipo numérico.

In [514]:
df["preco"] = (df["preco"]
    .str.extract(r"^R[$] (.*)")[0]
    .str.replace(".", "")
)

In [515]:
df["preco"].str.match(",").sum()

np.int64(0)

In [516]:
df["preco"] = df["preco"].astype("float")

### Feature `condominio`

In [517]:
get_info(df["condominio"])

dtype: str
NA: 39/7253 (0.54%)
valores:
               contagem porcentagem
condominio                         
não informado      1587     21.88 %
R$ 1                346      4.77 %
R$ 600              344      4.74 %
R$ 500              339      4.67 %
R$ 400              258      3.56 %
...                 ...         ...
R$ 1.032              1      0.01 %
R$ 544                1      0.01 %
R$ 533                1      0.01 %
R$ 432                1      0.01 %
R$ 2.049              1      0.01 %

[536 rows x 2 columns]


Assim como foi feito no target, vamos extrair o valor numérico dessa feature e converter ela para um tipo numérico.

In [518]:
df["condominio"] = (df["condominio"]
    .str.extract(r"^R[$] (.*)")[0]
    .str.replace(".", "")
)

In [519]:
df["condominio"].str.match(",").sum()

np.int64(0)

In [520]:
df["condominio"] = (df["condominio"]
    .replace("não informado", pd.NA)
    .astype("float")
)

### Feature `iptu`

In [521]:
get_info(df["iptu"])

dtype: str
NA: 39/7253 (0.54%)
valores:
               contagem porcentagem
iptu                               
não informado      2956     40.76 %
R$ 1/ano            691      9.53 %
R$ 1.000/ano        213      2.94 %
R$ 500/ano          189      2.61 %
R$ 800/ano          175      2.41 %
...                 ...         ...
R$ 287/ano            1      0.01 %
R$ 1.750/ano          1      0.01 %
R$ 9.840/ano          1      0.01 %
R$ 5.966/ano          1      0.01 %
R$ 840/ano            1      0.01 %

[655 rows x 2 columns]


Façamos o mesmo que foi feito nas duas últimas colunas tratadas.

In [522]:
df["iptu"] = (df["iptu"]
    .str.extract(r"^R[$] (.*)/ano")[0]
    .str.replace(".", "")
)

In [523]:
df["iptu"].str.match(",").sum()

np.int64(0)

In [524]:
df["iptu"] = (df["iptu"]
    .replace("não informado", pd.NA)
    .astype("float")
)

### Feature `endereco`

In [525]:
get_info(df["endereco"])

dtype: str
NA: 0/7253 (0.00%)
valores:
                                                    contagem porcentagem
endereco                                                                
Setor Bueno, Goiânia - GO                                131      1.81 %
Setor Oeste, Goiânia - GO                                 61      0.84 %
Setor Marista, Goiânia - GO                               51      0.70 %
Parque Amazônia, Goiânia - GO                             37      0.51 %
Rua 56, 79 - Jardim Goiás, Goiânia - GO                   29      0.40 %
...                                                      ...         ...
Rua VV 8, 785 - Village Veneza, Goiânia - GO               1      0.01 %
Avenida G, 100 - Setor Oeste, Goiânia - GO                 1      0.01 %
Rua C160, 349 - Jardim América, Goiânia - GO               1      0.01 %
Avenida Circular, 751 - Eldorado Parque, Goiâni...         1      0.01 %
Rua 147, 100 - Setor Marista, Goiânia - GO                 1      0.01 %

[3878 rows 

Vamos apenas converter o tipo dessa feature para texto.

In [526]:
df["endereco"] = df["endereco"].astype("str")

### Feature `tamanho`

In [527]:
get_info(df["tamanho"])

dtype: str
NA: 0/7253 (0.00%)
valores:
         contagem porcentagem
tamanho                      
64 m²         160      2.21 %
76 m²         129      1.78 %
58 m²         127      1.75 %
74 m²         120      1.65 %
68 m²         115      1.59 %
...           ...         ...
322 m²          1      0.01 %
319 m²          1      0.01 %
231 m²          1      0.01 %
375 m²          1      0.01 %
188 m²          1      0.01 %

[348 rows x 2 columns]


Vamos checar se todos os registros seguem esse padrão de apresentação do tamanho do imóvel.

In [528]:
df[~df["tamanho"].str.match(r"\d+ m²")]["tamanho"]

1      80 - 100 m²
8       59 - 79 m²
116     42 - 61 m²
188     78 - 86 m²
245     77 - 81 m²
274     70 - 74 m²
310     43 - 61 m²
375     57 - 59 m²
505     85 - 88 m²
Name: tamanho, dtype: str

Como podemos ver, alguns registros têm uma faixa de valores do imóvel ao invés de um valor em si. Como são poucos registros, iremos manter apenas o maior valor dessas faixar de valores para obtermos um valor só.

In [529]:
df["tamanho"] = (df["tamanho"]
    .str.replace(r"^\d+ - ", "", regex=True)
    .str.extract(r"^(.*) m²")[0]
    .str.replace(r"\.", "")
    .astype("float")
)

### Feature `quartos`

In [530]:
get_info(df["quartos"])

dtype: str
NA: 1/7253 (0.01%)
valores:
               contagem porcentagem
quartos                            
3 quartos          4012     55.32 %
2 quartos          2125     29.30 %
4 quartos           795     10.96 %
1 quarto            250      3.45 %
5 quartos            55      0.76 %
6 quartos             7      0.10 %
2 - 3 quartos         5      0.07 %
NaN                   1      0.01 %
10 quartos            1      0.01 %
7 quartos             1      0.01 %
11 quartos            1      0.01 %


Vamos extrair a quantidade de quartos a partir do padrão acima.

In [531]:
df["quartos"] = (df["quartos"]
    .str.replace(r"^\d+ - ", "", regex=True)
    .str.extract(r"^(.*) quarto[s]?")[0]
    .astype("float")
)

### Feature `banheiros`

In [532]:
get_info(df["banheiros"])

dtype: str
NA: 1/7253 (0.01%)
valores:
                 contagem porcentagem
banheiros                            
2 banheiros          2539     35.01 %
4 banheiros          1623     22.38 %
3 banheiros          1087     14.99 %
1 banheiro            966     13.32 %
5 banheiros           784     10.81 %
6 banheiros           202      2.79 %
7 banheiros            35      0.48 %
9 banheiros             5      0.07 %
11 banheiros            4      0.06 %
2 - 3 banheiros         2      0.03 %
1 - 2 banheiros         2      0.03 %
8 banheiros             2      0.03 %
NaN                     1      0.01 %
10 banheiros            1      0.01 %


Façamos o mesmo que foi feito na última feature.

In [533]:
df["banheiros"] = (df["banheiros"]
    .str.replace(r"^\d+ - ", "", regex=True)
    .str.extract(r"^(.*) banheiro[s]?")[0]
    .astype("float")
)

### Feature `vagas_estacionamento`

In [534]:
get_info(df["vagas_estacionamento"])

dtype: str
NA: 202/7253 (2.79%)
valores:
                      contagem porcentagem
vagas_estacionamento                      
2 vagas                   3227     44.49 %
1 vaga                    2706     37.31 %
3 vagas                    821     11.32 %
4 vagas                    207      2.85 %
NaN                        202      2.79 %
5 vagas                     30      0.41 %
1 - 2 vagas                 20      0.28 %
6 vagas                     20      0.28 %
1 - 3 vagas                  3      0.04 %
7 vagas                      3      0.04 %
1 - 4 vagas                  2      0.03 %
8 vagas                      2      0.03 %
9 vagas                      2      0.03 %
12 vagas                     2      0.03 %
10 vagas                     2      0.03 %
16 vagas                     1      0.01 %
17 vagas                     1      0.01 %
20 vagas                     1      0.01 %
13 vagas                     1      0.01 %


E novamente respetimos o processo.

In [535]:
df["vagas_estacionamento"] = (df["vagas_estacionamento"]
    .str.replace(r"^\d+ - ", "", regex=True)
    .str.extract(r"^(.*) vaga[s]?")[0]
    .astype("float")
)

### Feature `andar`

In [536]:
get_info(df["andar"])

dtype: str
NA: 3991/7253 (55.03%)
valores:
           contagem porcentagem
andar                          
NaN            3991     55.03 %
2 andar         212      2.92 %
3 andar         208      2.87 %
5 andar         207      2.85 %
6 andar         194      2.67 %
4 andar         173      2.39 %
10 andar        153      2.11 %
8 andar         145      2.00 %
9 andar         143      1.97 %
7 andar         139      1.92 %
1 andar         124      1.71 %
16 andar        113      1.56 %
12 andar        109      1.50 %
18 andar        101      1.39 %
11 andar        100      1.38 %
14 andar         98      1.35 %
20 andar         98      1.35 %
15 andar         96      1.32 %
17 andar         94      1.30 %
22 andar         76      1.05 %
13 andar         75      1.03 %
19 andar         62      0.85 %
23 andar         62      0.85 %
21 andar         57      0.79 %
26 andar         53      0.73 %
27 andar         47      0.65 %
25 andar         42      0.58 %
24 andar         39      0.54

Vamos extrair o valor do andar a partir do padrão acima.

In [537]:
df["andar"] = (df["andar"]
    .str.replace(r"^\d+ - ", "", regex=True)
    .str.extract(r"^(.*) andar?")[0]
    .astype("float")
)

### Feature `descricao`

In [538]:
get_info(df["descricao"])

dtype: str
NA: 0/7253 (0.00%)
valores:
                                                    contagem porcentagem
descricao                                                               
Vendo Espaçoso Apartamento de 4 Quartos com Áre...         7      0.10 %
Apartamento Versátil e Confortável à Venda no R...         7      0.10 %
Excelente apartamento no Setor Parque Amazônia,...         6      0.08 %
Seu Refúgio Urbano no Residencial Malheiros, Ja...         6      0.08 %
A NEWINC Incorporadora acaba de lançar no Setor...         5      0.07 %
...                                                      ...         ...
Condomínio harmonia\r\nParque Oeste industrial...         1      0.01 %
O apartamento está localizado no bairro Residen...         1      0.01 %
Elev Marista com 158m², elevador privativo, and...         1      0.01 %
Residencial Cristal do Sudoeste - 89 m² - Setor...         1      0.01 %
Localizado em um dos bairros mais nobres de Goi...         1      0.01 %

[6916 rows 

Como não teremos acesso à descrição do imóvel durante a etapa de inferência, então iremos remover essa feature do nosso dataset.

In [539]:
df = df.drop("descricao", axis="columns")

### Features Binárias

Agora vamos analisar as features binárias.

In [540]:
binary_features = [
    "piscina",
    "elevador",
    "churrasqueira",
    "condominio_fechado",
    "academia",
    "espaco_gourmet",
    "jardim",
    "parque_infantil",
    "salao_festas",
    "lavanderia",
    "acesso_pcd",
    "espaco_coworking",
    "portao_eletronica",
    "sauna",
    "spa",
    "permitido_animais",
    "ar_condicionado",
    "varanda_gourmet",
    "quadra_poliesportiva",
    "bicicletario",
    "portaria_24h",
    "cozinha_americana",
    "sacada",
    "interfone",
    "area_servico",
    "cozinha_grande",
    "mobiliado",
    "cozinha",
    "salao_jogos",
    "deposito",
    "porteiro",
    "seguranca_24h",
    "armario_embutido",
    "internet",
    "armario",
    "cozinha_gourmet",
    "gerador_eletrico",
    "alarmes",
    "brinquedoteca",
    "cinema",
    "aquecimento",
    "moveis_embutidos",
    "armarios_cozinha",
    "circuito_seguranca",
    "garagem",
    "quadra_tenis",
    "area_verde",
    "squash",
    "lareira",
    "escritorio",
    "armarios_banheiro",
    "chuveiro_gas",
    "box_blindex",
    "recepcao",
    "tv_cabo",
    "banheira",
    "piscina_privativa",
    "quintal",
    "gramado",
    "janelas_grandes",
    "ventilacao_natural",
    "fogao",
    "trilha_caminhada",
    "vista_panoramica",
    "massagem",
    "freezer"
]

for feature in binary_features:
    unique = df[feature].unique()
    values = list(unique)
    
    print(f"{feature}: {values}")

piscina: ['Piscina', nan]
elevador: ['Elevador', nan]
churrasqueira: ['Churrasqueira', nan]
condominio_fechado: ['Condomínio fechado', nan]
academia: ['Academia', nan]
espaco_gourmet: ['Espaço gourmet', nan]
jardim: ['Jardim', nan]
parque_infantil: ['Playground', nan]
salao_festas: ['Salão de festas', nan]
lavanderia: ['Lavanderia', nan]
acesso_pcd: ["Acesso para PCD's", nan]
espaco_coworking: ['Coworking', nan]
portao_eletronica: ['Portão eletrônico', nan]
sauna: ['Sauna', nan]
spa: ['Spa', nan]
permitido_animais: [nan, 'Animais permitidos']
ar_condicionado: [nan, 'Ar condicionado']
varanda_gourmet: [nan, 'Varanda gourmet']
quadra_poliesportiva: [nan, 'Quadra poliesportiva']
bicicletario: [nan, 'Bicicletário']
portaria_24h: [nan, 'Portaria 24h']
cozinha_americana: [nan, 'Cozinha americana']
sacada: [nan, 'Varanda']
interfone: [nan, 'Interfone']
area_servico: [nan, 'Área de serviço']
cozinha_grande: [nan, 'Cozinha grande']
mobiliado: [nan, 'Mobiliado']
cozinha: [nan, 'Cozinha']
salao_j

Como podemos ver, todas as features são binárias, ou seja, ou o imóvel tem a característica descrita na feature ou não tem. Vamos converter essas features para o tipo booleano.

In [541]:
df[binary_features] = df[binary_features].notna()

## Convertendo os `dtypes`

Agora, antes de finalizar, iremos converter os `dtypes` do nosso dataset.

In [542]:
df = df.convert_dtypes()

## Salvando o Dataset

Para finalizar, vamos salvar nosso dataset limpo em formato parquet.

In [543]:
df.to_parquet("../data/processed/01_cleaned.parquet")